### ISSUE TITLE
XGBoost histogram counting to construct a tree (processing column in batches vs in one go)

### ISSUE THREAD

**Issue by VigneshN1997** [dmlc/xgboost#4906]
https://github.com/dmlc/xgboost/blob/d30e63a0a519d6b5ee2d30902eb8f64a1c68fcf8/src/tree/updater_histmaker.cc#L418

Hi everyone. I am working on the xgboost algorithm and I noticed in **updater_histmaker** that while proposing split points in **ResetPosAndPropose**, the **UpdateSketchCol** function was called on a row batch of a column (i.e. data is divided into batches and one particular column of a batch is passed to the **UpdateSketchCol**). But in xgboost 0.9 we arent getting the row batches and then calling **UpdateSketchCol**. Instead, we call **GetSortedColumnBatches** and pass that column completely to the UpdateSketchCol.

**Earlier: a particular column was passed in batches to UpdateSketchCol
Now: complete column passed in one go to UpdateSketchCol**

For example: suppose we have a feature value x appearing 100 times.
Earlier:
If we have two batches, in one batch x appears 80 times and in another batch x appears 20 times. 
Now: there will be only one batch with weight of x = 100 (wmin=100).  

So the weight of x (**wmin**) will be different in both cases and this will cause value of rmax to be different. If **rmax** is different, the **Push**() function in SketchEntry struct (in **updater_basemaker-inl.h**) will behave differently (since it pushes to sketch by comparing **rmax** and **next_goal**). This difference is causing different feature values to be added to the sketchs array which leads to different number of points being proposed (as candidate split points). 

I get it that the DMatrix structure was refactored completely. But why is data now being processed completely in one go now rather than in row batches?

I wanted to understand why this change happened in **updater histmaker**. It would be helpful for me if I could get the reason behind the change in histmaker (since it is causing the algorithm to work differently).

**Comment by trivialfis** [comment]
I haven't really look into histmaker and more familiar with `hist` and `gpu_hist`.

The term `batch` is only meaningful in external memory training (out of core).  If your data fits into memory, the batch is the whole dataset.

**Comment by VigneshN1997** [comment]
So if we compare the results from earlier (quantiles created in batches of size 2^15) and now (complete data fitting in memory), will the current implementation give me better results? If yes, why will it give the better results?  I am asking this because I wanted to understand the motivation behind the change in code.

**Comment by thvasilo** [comment]
Hello @VigneshN1997 I've been looking into updater_histmaker recently, but mostly on the .75 branch. Could you point to change that happened? I think I understand what you're talking about with data being fed per batch, but not sure when and why that changed.

**Comment by VigneshN1997** [comment]
The change was done in this commit: https://github.com/dmlc/xgboost/commit/70d208d68c3a32aaa4fcd6aa456f286a4da5912f#diff-cde48cc62edcbcd1eeeef1805d1f3f51

In this commit check out the learner.cc and updater_histmaker.cc files. In updater_histmaker.cc , quantiles are being formed on SortedColumnBatches (all rows present in a column), while earlier it was formed on row batches (each row batch had the column values sorted). In learner.cc, the call to InitColAccess and the attribute max_row_perbatch has been removed. The InitColAccess function was dividing up the data in batches of size max_row_perbatch (default value = 2^15). So after InitColAccess, when the updater was called, the data was processed in row batches.

**Comment by thvasilo** [comment]
I only took a brief look so I'm not sure I follow, I'll ping @RAMitchell who's the original author of this change.

Both before and after this PR, we iterate over columns fist (lines 426 and 711 in the linked PR) when accumulating statistics. This is done to accumulate histogram statistics,  in `CreateHist` (I'm focusing on `GlobalProposalHistMaker` since that's the one I'm familiar with).

Then there is a second iteration in `ResetPosAndPropose` where it's the only place where we iterate over rows.

As far as I can tell, we are still only passing a batch of columns, as the current version (which @rongou contributed) uses a templated call of `DMatrix::GetBatches` to subsequently call `[GetSortedColumnBatches](https://github.com/dmlc/xgboost/blob/125bcec62e2cc42de5ee97442c357000e9e5b1ff/include/xgboost/data.h#L486)`.

I was under the impression that previously we still had "batches of columns" but you're saying (and I'm sorry for repeating):

> In updater_histmaker.cc , quantiles are being formed on SortedColumnBatches (all rows present in a column), while earlier it was formed on row batches (each row batch had the column values sorted).

So you're saying that before https://github.com/dmlc/xgboost/pull/3395, we would take a batch of rows, and iterate over them sorted by the values of particular column. Now we take _all_ the rows instead, but column values are still sorted, correct?

**Comment by rongou** [comment]
I'm fairly confident my changes are pure refactorings, shouldn't have any behavior differences.

@VigneshN1997 is there a way for you to write a (failing) test to demonstrate the undesired behavior?

**Comment by VigneshN1997** [comment]
@thvasilo , yes, before #3395 we would take a batch of rows, and iterate over them sorted by the values of particular column. Now we take all the rows instead, but column values are still sorted.

@rongou this is the test which shows the updesirable behaviour:

Created a dataset (34000 records) with only one numerical column and one label column .

```
import csv
import random
rows = []
n = 34000 
zero_count = random.randint(1, n)
one_count = n - zero_count
labels = [0]*zero_count + [1]*one_count
random.shuffle(labels)
for i in range(n):
    rows.append([i,labels[i]]) # numerical feature (0-33999) and label
train_file = 'synthetic.csv'
with open(train_file, 'w') as f:
    f_writer = csv.writer(f, delimiter=',')
    for i in range(n):
        f_writer.writerow(rows[i])
```

Compared the tags v0.72 and v0.90:

In Push function of SketchEntry (in updater_basemaker-inl.h):
added this logging statement just before adding to sketch:
`LOG(CONSOLE) << "added to sketch:" << " rmin:" << rmin << " rmax:" << rmax << " wmin:" << wmin << " last_fvalue:" << last_fvalue;`
Line 411 (v0.72) , line 401 (v0.90):
```
            sketch->temp.data[sketch->temp.size] =
                common::WXQuantileSketch<bst_float, bst_float>::
                Entry(static_cast<bst_float>(rmin),
                      static_cast<bst_float>(rmax),
                      static_cast<bst_float>(wmin), last_fvalue);

```
added this logging statement before calling UpdateSketchCol:
Line 441 (v0.72) , line 433 (v0.90):
`LOG(CONSOLE) << "calling update sketch col on feature:" << batch.col_index[i] << " column size:" << batch[i].length;`
```
this->UpdateSketchCol(gpair, batch[fid], tree,
                                  work_set_size, offset,
                                  &thread_sketch_[omp_get_thread_num()]);

```

added this logging statement to print final cut points:
after Line 468 (v0.72) , after line 460 (v0.90):
`LOG(CONSOLE) << "i:" << i << " cpt:" << cpt;`
```
for (size_t i = 1; i < a.size; ++i) {
            bst_float cpt = a.data[i].value - kRtEps;
            LOG(CONSOLE) << "i:" << i << " cpt:" << cpt;
            if (i == 1 || cpt > this->wspace_.cut.back()) {
              this->wspace_.cut.push_back(cpt);
            }
          }

```

Command run:

```
$RND=path to xgboost executable
$CONF_FILE=path to empty config file

$RND $CONF_FILE \
 booster=gbtree \
 task=train \
 train_path=$DATA \
 nthread=1 \
 num_round=1 \
 max_depth=5 \
 learning_rate=0.1 \
 min_child_weight=1.0 \
 reg_alpha=0 \
 reg_lambda=0 \
 min_split_loss=0 \
 objective=binary:logistic \
 model_out=$MODEL \
 tree_method=approx \
 sketch_eps=0.1;

```

sketch_eps=0.1 in both cases

I am attaching the results. Observations from results:

- In v0.72, the data is processed in batches (32768+1232), in v0.90 data is processed completely in one pass (34000).
- The above mentioned point, causes different values to be pushed to the sketch.
- Thus, the final cut points proposed are different in both the cases.

[results_0.72.txt](https://github.com/dmlc/xgboost/files/3689101/results_0.72.txt)
[results_0.90.txt](https://github.com/dmlc/xgboost/files/3689102/results_0.90.txt)

**Comment by thvasilo** [comment]
I have not tested this myself but this indeed looks like it shouldn't be happening. The question is which behavior is correct.

I think that as long as all the rows can fit into memory, it's better now to create the split point according to the statistics from all data points.

**Comment by rongou** [comment]
Just looking at the results, seems like 0.90 is more correct since the cuts are evenly divided, no?

**Comment by thvasilo** [comment]
Correct that should be the expected behavior. @hcho3 can verify this.

**Comment by VigneshN1997** [comment]
@rongou This behaviour could be because this is the first boosting round and all records have the same hessian value (0.25) and the data is also uniform (each feature 0-33999 in steps of 1). So yes, the cuts should be evenly divided. But can something be proved theoretically or improvement in performance metric that could justify the design choice (of processing complete data in memory)?

**Comment by rongou** [comment]
@RAMitchell would have to chime in for the rationale behind this change. To me this seems ok, the results are more accurate, probably more efficient too since we are dealing with one column at a time.

**Comment by VigneshN1997** [comment]
@rongou , @RAMitchell , @thvasilo any update on the rationale behind this change?
